Userland/Libraries/LibAudio/QOALoader.cpp

```diff
@@ -0,0 +1,258 @@
+        is_first_frame = IsFirstFrame::No;
+        VERIFY(slice_to_load_into.size() <= QOA::max_frame_samples);
+        current_loaded_samples += slice_to_load_into.size();
+        // The buffer wasn't large enough for the next frame.
+        if (slice_to_load_into.size() == 0)
+            break;
+    }
+    m_loaded_samples += current_loaded_samples;
+    auto trimmed_samples = LOADER_TRY(FixedArray<Sample>::create(samples.span().trim(current_loaded_samples)));
+
+    return trimmed_samples;
+}
+
+MaybeLoaderError QOALoaderPlugin::reset()
+{
+    LOADER_TRY(m_stream->seek(QOA::header_size, AK::SeekMode::SetPosition));
+    m_loaded_samples = 0;
+    // Read the first frame, then seek back to the beginning. This is necessary since the first frame contains the sample rate and channel count.
+    auto frame_samples = LOADER_TRY(FixedArray<Sample>::create(QOA::max_frame_samples));
+    auto span = frame_samples.span();
+    LOADER_TRY(load_one_frame(span, IsFirstFrame::Yes));
+
+    LOADER_TRY(m_stream->seek(QOA::header_size, AK::SeekMode::SetPosition));
+    m_loaded_samples = 0;
+    return {};
+}
+
+MaybeLoaderError QOALoaderPlugin::seek([[maybe_unused]] int sample_index)
+{
+    if (sample_index == 0 && m_loaded_samples == 0)
+        return {};
+    // A QOA file consists of 8 bytes header followed by a number of usually fixed-size frames.
+    // This fixed bitrate allows us to seek in constant time.
+    if (!m_has_uniform_channel_count)
+        return LoaderError { LoaderError::Category::Unimplemented, LOADER_TRY(m_stream->tell()), "QOA with non-uniform channel count is currently not seekable"sv };
+    /// FIXME: Change the Loader API to use size_t.
+    VERIFY(sample_index >= 0);
+    // We seek to the frame "before"; i.e. the frame that contains that sample.
+    auto const frame_of_sample = static_cast<size_t>(AK::floor<double>(static_cast<double>(sample_index) / static_cast<double>(QOA::max_frame_samples)));
+    auto const frame_size = QOA::frame_header_size + m_num_channels * (QOA::lms_state_size + sizeof(QOA::PackedSlice) * QOA::max_slices_per_frame);
+    auto const byte_index = QOA::header_size + frame_of_sample * frame_size;
+    LOADER_TRY(m_stream->seek(byte_index, AK::SeekMode::SetPosition));
+    m_loaded_samples = frame_of_sample * QOA::max_frame_samples;
+    return {};
+}
+
+MaybeLoaderError QOALoaderPlugin::read_one_slice(QOA::LMSState& lms_state, Span<i16>& samples)
+{
+    VERIFY(samples.size() == QOA::slice_samples);
+
+    auto packed_slice = LOADER_TRY(m_stream->read_value<BigEndian<u64>>());
+    auto unpacked_slice = unpack_slice(packed_slice);
+
+    for (size_t i = 0; i < QOA::slice_samples; ++i) {
+        auto const residual = unpacked_slice.residuals[i];
+        auto const predicted = lms_state.predict();
+        auto const dequantized = QOA::dequantization_table[unpacked_slice.scale_factor_index][residual];
+        auto const reconstructed = clamp(predicted + dequantized, -32768, 32767);
```

Comment:
Maybe use constants for `-32768` and `32767`?
Just to have some names mapping to spec terms / with a comment like `The result is clamped to int16 to form the final output sample.`

Reply:
Sure, I was *very much* porting the reference implementation here.